### PR TITLE
move replicas action functionality into AllocateAction

### PR DESCRIPTION
**Pull request by talevy**
Since replica counts and allocation rules are set separately, it is not always clear how many replicas are to be allocated in the allocate action. Moving the replicas action to occur at the same time as the allocate action, resolves this confusion that could end an undesired state. This means that the ReplicasAction is removed, and a new optional replicas parameter is added to AllocateAction.

### CODE REVIEW

File: x-pack/plugin/ilm/src/test/java/org/elasticsearch/xpack/indexlifecycle/TimeSeriesLifecycleActionsIT.java [hunk 1]
```diff
@@ -71,6 +70,22 @@ public void testAllocate() throws Exception {
...
+            .put(IndexMetaData.SETTING_NUMBER_OF_REPLICAS, numReplicas));
+        AllocateAction allocateAction = new AllocateAction(finalNumReplicas, null, null, null);
+        createNewSingletonPolicy(randomFrom("warm", "cold"), allocateAction);
+        updatePolicy(index, policy);
+        assertBusy(() -> {
+            Map<String, Object> settings = getOnlyIndexSettings(index);
+            assertThat(getStepKey(settings), equalTo(TerminalPolicyStep.KEY));
+            assertThat(settings.get(IndexMetaData.INDEX_NUMBER_OF_REPLICAS_SETTING.getKey()), equalTo(String.valueOf(finalNumReplicas)));
+        });
+    }
```
Comment (colings86) [inline x-pack/plugin/ilm/src/test/java/org/elasticsearch/xpack/indexlifecycle/TimeSeriesLifecycleActionsIT.java]: Could we also add a test that tests the bug I found (so where we change both replicas and allocate at the same times on a configuration where the resulting nodes would not be enough to allocate the original number of replicas)?

**Review by colings86 — CHANGES_REQUESTED** [review]
I left a comment asking for an extra test but otherwise I think this is good

**Comment by talevy** [comment]
@colings86 thanks for taking a look. I've added a unit test to check the allocation of an unallocated replica

**Review by colings86 — APPROVED** [review]
LGTM

**Pull request merged**